lib/SDIR/Translate/translateToSDFG.cpp

```diff
@@ -0,0 +1,1597 @@
+//===----------------------------------------------------------------------===//
+
+LogicalResult printConstant(arith::ConstantOp &op, JsonEmitter &jemit) {
+  Type t = op.getType();
+  std::string val;
+  llvm::raw_string_ostream valStream(val);
+  op.getValue().print(valStream);
+  val.erase(val.find(' '));
+
+  AsmState state(op);
+  std::string res;
+  llvm::raw_string_ostream resStream(res);
+  op.getResult().printAsOperand(resStream, state);
+
+  jemit.startNamedList(res);
+
+  jemit.startObject();
+  jemit.printKVPair("type", "Scalar");
+
+  jemit.startNamedObject("attributes");
+  jemit.printKVPair("allow_conflicts", "false", /*stringify=*/false);
+  translateTypeToSDFG(t, op.getLoc(), jemit, "dtype");
+  jemit.startNamedList("shape");
+  jemit.printString("1");
+  jemit.endList(); // shape
+  jemit.printKVPair("transient", "false", /*stringify=*/false);
+  jemit.endObject(); // attributes
+
+  jemit.endObject();
+
+  jemit.startEntry();
+  jemit.printLiteral(val);
+  jemit.endList(); // res
+
+  return success();
+}
+
+LogicalResult printSDFGNode(SDFGNode &op, JsonEmitter &jemit) {
+  jemit.printKVPair("type", "SDFG");
+  jemit.printKVPair("sdfg_list_id", SDIRDialect::getNextID(),
+                    /*stringify=*/false);
+
+  jemit.startNamedObject("attributes");
+  jemit.printAttributes(
+      op->getAttrs(),
+      /*elidedAttrs=*/{"ID", "entry", "sym_name", "type", "arg_names"});
+
+  jemit.startNamedObject("constants_prop");
+  for (Operation &oper : op.body().getOps())
+    if (StateNode state = dyn_cast<StateNode>(oper))
+      for (arith::ConstantOp constOp : state.getConstants())
+        if (printConstant(constOp, jemit).failed())
+          return failure();
+  jemit.endObject(); // constants_prop
+
+  if (containsAttr(*op, "arg_names")) {
+    Attribute arg_names = op->getAttr("arg_names");
+    if (ArrayAttr arg_names_arr = arg_names.cast<ArrayAttr>()) {
+      jemit.startNamedList("arg_names");
+
+      for (Attribute arg_name : arg_names_arr.getValue()) {
+        if (StringAttr arg_name_str = arg_name.cast<StringAttr>()) {
+          jemit.startEntry();
+          jemit.printString(arg_name_str.getValue());
+        } else {
+          mlir::emitError(op.getLoc(),
+                          "'arg_names' must consist of StringAttr");
+          return failure();
+        }
+      }
+
+      jemit.endList(); // arg_names
+    } else {
+      mlir::emitError(op.getLoc(), "'arg_names' must be an ArrayAttr");
+      return failure();
+    }
+  } else {
+    jemit.startNamedList("arg_names");
+
+    for (BlockArgument bArg : op.getArguments()) {
+      AsmState state(op);
+      std::string name;
+      llvm::raw_string_ostream nameStream(name);
+      bArg.printAsOperand(nameStream, state);
+      name.erase(0, 1); // Remove %-sign
+      jemit.startEntry();
+      jemit.printString(name);
+    }
+
+    jemit.endList(); // arg_names
+  }
+
+  jemit.startNamedObject("_arrays");
+  for (Operation &oper : op.body().getOps()) {
```

Comment:
didn't you implement methods that get all allocations in Ops?

Reply:
Not for `SDFGNode` but for `StateNode`. Should I add those helper function to condense the code a bit?

Reply:
They should be accessible. Maybe as part of some general SDIR utils / abstract class?

Reply:
Nvm. MLIR already has a function for this